libarchive/archive_read_disk_posix.c

```diff
@@ -1420,7 +1420,7 @@ update_current_filesystem(struct archive_read_disk *a, int64_t dev)
 	 */
 	fid = t->max_filesystem_id++;
 	if (t->max_filesystem_id > t->allocated_filesystem) {
-		size_t s;
+		int s;
```

Comment:
This looks okay.  It would be nice while you're here to harden this code a bit:  Please add a check just above that returns a fatal error if `fid` becomes unreasonably large.  That code might look like this:
```
if (fid > MAX_FILESYSTEM_ID) {
  archive_set_error(&a->archive, ENOMEM, "Too many filesystems");
  return (ARCHIVE_FATAL);
}
```
and then `#define MAX_FILESYSTEM_ID 1000000` to provide an upper bound.   In particular, this will help ensure that bugs in other areas don't lead us to accidentally overflow `s` here.

Also, note that very similar code appears in `archive_read_disk_windows.c` -- that needs the same attention, I presume.